[WIP] AWS: Allow credentials to be manually provided

jeana-redhat opened the pull request.

jeana-redhat commented:
Jira link: https://issues.redhat.com/browse/OSDOCS-1231
Note: this covers CO-925 and CO-981

jeana-redhat commented:
@dgoodwin & @joelddiaz PTAL at this content for manual mode. CO-925 mentions AWS specifically, but the content is shared with Azure and GCP.

* [AWS](https://co-925--ocpdocs.netlify.app/openshift-enterprise/latest/installing/installing_aws/manually-creating-iam.html)
* [Azure](https://co-925--ocpdocs.netlify.app/openshift-enterprise/latest/installing/installing_azure/manually-creating-iam-azure.html)
* [GCP](https://co-925--ocpdocs.netlify.app/openshift-enterprise/latest/installing/installing_gcp/manually-creating-iam-gcp.html)
* [Upgrade pre-req step](https://co-925--ocpdocs.netlify.app/openshift-enterprise/latest/updating/updating-cluster-between-minor.html#prerequisites) (this is duplicated across all upgrade topics)

joelddiaz commented:
for AWS: reorder the steps so that what is now step 1 becomes step 3. So new order (based on current numbers): 2, 3, 1,

joelddiaz commented on modules/manually-maintained-credentials-upgrade.adoc:
```diff
@@ -5,9 +5,22 @@
-situations where a user might enter an upgrade that will fail because their
-manually maintained credentials have not been updated to match the
-`CredentialsRequests` in the upcoming release image.
+If permissions change in a future release, the Cloud Credential Operator (CCO)
```
@dgoodwin how best to document this? If there are brand new credentials, CCO can mark itself upgradable=False (all that work you did in those recent PRs), but CCO doesn't really have a way to handle if ExistingCredReqA has new permissions requirements from one version to the next (eg. image-registry now needs some encryption-related permissions that weren't previously required). The best we can presently do is hope that the secret containing the credentials is good enough. Given the limitations with permissions simulations, I'm reluctant to rely on the results of permissions simulations solely to be able to reliably be able to flag "ExistingCredReqA needs updating".

jeana-redhat replied:
@dgoodwin & @joelddiaz  -
Circling back on this: with the step order fixed, I think this comment is the only outstanding issue on this PR. LMK what y'all think we ought to say about it and I'll be happy to get this one wrapped up.

dgoodwin replied:
This is an interesting one, in manual mode we don't do *any* of the policy checking to make sure it's in sync with what's in the CredentialsRequest IIRC. We still should have our read-only cred, so in theory we could change that and start comparing policy and reporting problems.

If we did do that work, then we would want to update the upgradeable false to not just look for existence of a secret but also try to verify the permissions are correct on it. 

So today just a secret check, which is consistent with manual mode behavior anyhow, it's on you to make sure the permissions are right. We could possibly improve this in future.

dgoodwin replied:
No policy simulation required I hope.

dgoodwin replied:
So I guess we need another section here, something like:

In addition to creating new credentials for the release image you wish to upgrade to, you must also compare the required permissions for existing credentials to accommodate any pre-existing components which require new permissions.

This would flow into the last line about seeing the manual IAM instructions.

jeana-redhat commented:
> for AWS: reorder the steps so that what is now step 1 becomes step 3. So new order (based on current numbers): 2, 3, 1,

Just to be clear - only for AWS and _not_ for Azure or GCP?

joelddiaz commented:
> Just to be clear - only for AWS and _not_ for Azure or GCP?
All of them should be step1: create install-config; step2: edit install-config; step3: generate manifests

jeana-redhat commented:
force push just now addresses out of order steps - should now be correct

jeana-redhat commented:
Updates today:

- added explanation that users need to check existing creds before upgrading to make sure the permissions haven't changed in the new release
- used conditions to show credentialsmode=manual for AWS while leaving Azure and GCP with the configmap override

@dgoodwin & @joelddiaz PTAL

dgoodwin commented on modules/manually-maintained-credentials-upgrade.adoc:
```diff
@@ -5,9 +5,28 @@
+If permissions change in a future release, the Cloud Credential Operator (CCO)
+`upgradable` status for a cluster with manually maintained credentials changes
+to `false`. For minor releases (for example, from 4.5 to 4.6), this status
+prevents you from upgrading until you have addressed any updated permissions.
```
Possible nit here, this forces you to create Secrets for any new credentials. We don't/can't actually check yet for new permissions being added or removed from existing credentials. We might be able to improve this in the future but it could be tricky.

jeana-redhat replied:
I added lines 25-29 to cover the new permissions for existing creds scenario. So would it work to change the first line here as follows and then leave the rest of the section as-is?? 
`If credentials are added in a future release,`

dgoodwin replied:
:+1:

jeana-redhat commented:
I believe everything discussed here is resolved, so I am going to move this on to QE review. Thanks!

jeana-redhat commented:
@lwan-wanglin I believe you're the QE contact for this - can you PTAL?

lwan-wanglin commented:
@yunjiang29 Could you please help to double review this doc PR?  this is for https://issues.redhat.com/browse/CO-925.

jeana-redhat commented on modules/manually-create-identity-access-management.adoc:
```diff
@@ -188,8 +221,6 @@ $ openshift-install create cluster --dir=mycluster
 +
```
Very good catch! Fixed.

jeana-redhat commented on modules/manually-create-identity-access-management.adoc:
```diff
@@ -15,22 +15,55 @@ ifeval::["{context}" == "manually-creating-iam-gcp"]
+. Edit the `install-config.yaml` configuration file so that the CCO operates in
+manual mode:
++
+[source,terminal]
+----
+$ oc edit cloudcredentials.operator.openshift.io cluster
+----
```
That step only shows up for AWS (Azure and GCP use the old way, after generating manifests), does that seem right? Or still not right?

yunjiang29 replied:
@jeana-redhat 
1. for you question:

    > That step only shows up for AWS (Azure and GCP use the old way, after generating manifests), does that seem right? Or still not right?

    right, `credentialsMode: Manual `is only effective AWS. 

2. since at this time the cluster has not been installed yet, so we can not use `oc edit cloudcredentials.operator.openshift.io cluster` command. 
To add `credentialsMode: Manual` option, just need to edit `install-config.yaml` file, e.g. `vi <installation_directory>/install-config.yaml`

jeana-redhat replied:
Thanks for clarifying :)

yunjiang29 commented:
@jeana-redhat some comments I can not add it to the source code, I list them here, sorry for the inconvenience.
1. in section [1] , step 1, step 3 and step 5, there are 3 different way to invoke `openshift-install` command:
`./openshift-install`, `openshift-install`, `bin/openshift-install`, can we only use `openshift-install`, the same as  `oc` command, this will have a good user experience 

2. in section [1] , step 6, I do not think we need this step, since in step 7, we listed all CRs for AWS, this step can be removed.

3. in section [1] , step 7
* command should be `oc adm release extract quay.io/openshift-release-dev/ocp-release:4.z.z-x86_64 --credentials-requests --cloud=aws`
* following scentence can be removed, since we listed all AWS related CRs, there is no other CRs here.

>  Remember to ignore any CredentialsRequests where the spec.providerSpec.kind does not match the cloud provider you are installing to.

* title could change to `Locate all CredentialsRequests in this release image that target the cloud you are deploying on:`

4. in section [1] , step 8
 change `request.spec.secretRef` to `spec.secretRef` ?

[1] https://co-925--ocpdocs.netlify.app/openshift-enterprise/latest/installing/installing_aws/manually-creating-iam.html#manually-create-iam_manually-creating-iam-aws

yunjiang29 commented:
@jeana-redhat For Azure platform, silmilar as AWS platform:

1. in section [1] , step 4, change `bin/openshift-install version` to `openshift-install version` ?

2. in section [1] , step 5, this step can be removed, since we have listed all CRs in step 6

3. in section [1] , step 6
* command should be `oc adm release extract quay.io/openshift-release-dev/ocp-release:4.z.z-x86_64 --credentials-requests --cloud=azure`
* following scentence can be removed:

>  Remember to ignore any CredentialsRequests where the spec.providerSpec.kind does not match the cloud provider you are installing to.

* title could change to `Locate all CredentialsRequests in this release image that target the cloud you are deploying on:`

4. in section [1] , step 7
 change `request.spec.secretRef` to `spec.secretRef` ?

[1] https://co-925--ocpdocs.netlify.app/openshift-enterprise/latest/installing/installing_azure/manually-creating-iam-azure.html#manually-create-iam_manually-creating-iam-azure

yunjiang29 commented:
@jeana-redhat For GCP platform, silmilar as AWS platform:

1. in section [1] , step 4, change `bin/openshift-install version` to `openshift-install version` ?

2. in section [1] , step 5, this step can be removed, since we have listed all CRs in step 6

3. in section [1] , step 6
* command should be `oc adm release extract quay.io/openshift-release-dev/ocp-release:4.z.z-x86_64 --credentials-requests --cloud=gcp`
* following scentence can be removed:

>  Remember to ignore any CredentialsRequests where the spec.providerSpec.kind does not match the cloud provider you are installing to.

* title could change to `Locate all CredentialsRequests in this release image that target the cloud you are deploying on:`

4. in section [1] , step 7
 change `request.spec.secretRef` to `spec.secretRef` ?

[1] https://co-925--ocpdocs.netlify.app/openshift-enterprise/latest/installing/installing_gcp/manually-creating-iam-gcp.html#manually-create-iam_manually-creating-iam-gcp

jeana-redhat commented:
@yunjiang29 Thanks so much for the detailed and careful feedback - the way you presented it was not inconvenient at all :) I believe I have addressed all issues, and would appreciate it if you (and @joelddiaz ) would give it a check. 

Note: I'm not sure how long it takes the preview-bot to generate the output, it's not up-to-date at the time of this comment but should be soon.

joelddiaz reviewed (commented):
:+1:

yunjiang29 commented:
@jeana-redhat thanks for update, LGTM

jeana-redhat commented:
@openshift/team-documentation ready for peer review :)

bmcelvee commented on modules/manually-maintained-credentials-upgrade.adoc:
```diff
@@ -5,9 +5,28 @@
-`CredentialsRequests` in the upcoming release image.
+If credentials are added in a future release, the Cloud Credential Operator
+(CCO) `upgradable` status for a cluster with manually maintained credentials
+changes to `false`. For minor releases (for example, from 4.5 to 4.6), this
```
The ISG style guide is fairly strict on using parentheses. If there's a way to re-write the parenthetical, that would be ideal.

jeana-redhat replied:
I could do it with commas instead, but thought that was a bit unwieldy. But if a bit unwieldy is preferred over parens, I'll do it ;)

bmcelvee commented on modules/manually-maintained-credentials-upgrade.adoc:
```diff
@@ -5,9 +5,28 @@
+upgrade is not blocked, but the credentials must still be updated for the new
+release.
+
+To determine if the CCO is upgradeable, switch to the *Administrator*
```
Could you make this bit a procedure?

jeana-redhat replied:
Good call :)

bmcelvee commented on updating/updating-cluster-between-minor.adoc:
```diff
@@ -18,6 +18,7 @@ Because of the difficulty of changing update channels by using `oc`, use the web
 See xref:../authentication/using-rbac.adoc[Using RBAC to define and apply permissions].
 * Have a recent xref:../backup_and_restore/backing-up-etcd.adoc#backup-etcd[etcd backup] in case your upgrade fails and you must xref:../backup_and_restore/disaster_recovery/scenario-2-restoring-cluster-state.adoc#dr-restoring-cluster-state[restore your cluster to a previous state].
 * Ensure all Operators previously installed through Operator Lifecycle Manager (OLM) are updated to their latest version in their latest channel. Updating the Operators ensures they have a valid upgrade path when the default OperatorHub catalogs switch from the current minor version to the next during a cluster upgrade. See xref:../operators/admin/olm-upgrading-operators.adoc#olm-upgrading-operat ...
+* If your cluster uses manually maintained credentials, ensure that the CCO is in an upgradeable state. For more information, see _Upgrading clusters with manually maintained credentials_ for xref:../installing/installing_aws/manually-creating-iam.adoc#manually-maintained-credentials-upgrade_manually-creating-iam-aws[AWS], xref:../installing/installing_azure/manually-creating-iam-azure.adoc#manua ...
```
s/CCO/Cloud Credential Operator (CCO)

jeana-redhat replied:
oh yeah whoops :+1:

bmcelvee commented on updating/updating-cluster-cli.adoc:
```diff
@@ -13,6 +13,7 @@ You can update, or upgrade, an {product-title} cluster within a minor version by
 See xref:../authentication/using-rbac.adoc[Using RBAC to define and apply permissions].
 * Have a recent xref:../backup_and_restore/backing-up-etcd.adoc#backup-etcd[etcd backup] in case your upgrade fails and you must xref:../backup_and_restore/disaster_recovery/scenario-2-restoring-cluster-state.adoc#dr-restoring-cluster-state[restore your cluster to a previous state].
 * Ensure all Operators previously installed through Operator Lifecycle Manager (OLM) are updated to their latest version in their latest channel. Updating the Operators ensures they have a valid upgrade path when the default OperatorHub catalogs switch from the current minor version to the next during a cluster upgrade. See xref:../operators/admin/olm-upgrading-operators.adoc#olm-upgrading-operat ...
+* If your cluster uses manually maintained credentials, ensure that the CCO is in an upgradeable state. For more information, see _Upgrading clusters with manually maintained credentials_ for xref:../installing/installing_aws/manually-creating-iam.adoc#manually-maintained-credentials-upgrade_manually-creating-iam-aws[AWS], xref:../installing/installing_azure/manually-creating-iam-azure.adoc#manua ...
```
s/CCO/Cloud Credential Operator (CCO)

bmcelvee commented on updating/updating-cluster-rhel-compute.adoc:
```diff
@@ -14,6 +14,7 @@ those machines.
 * Have access to the cluster as a user with `admin` privileges.
 See xref:../authentication/using-rbac.adoc[Using RBAC to define and apply permissions].
 * Have a recent xref:../backup_and_restore/backing-up-etcd.adoc#backup-etcd[etcd backup] in case your upgrade fails and you must xref:../backup_and_restore/disaster_recovery/scenario-2-restoring-cluster-state.adoc#dr-restoring-cluster-state[restore your cluster to a previous state].
+* If your cluster uses manually maintained credentials, ensure that the CCO is in an upgradeable state. For more information, see _Upgrading clusters with manually maintained credentials_ for xref:../installing/installing_aws/manually-creating-iam.adoc#manually-maintained-credentials-upgrade_manually-creating-iam-aws[AWS], xref:../installing/installing_azure/manually-creating-iam-azure.adoc#manua ...
```
s/CCO/Cloud Credential Operator (CCO)

bmcelvee commented on updating/updating-cluster.adoc:
```diff
@@ -12,6 +12,7 @@ You can update, or upgrade, an {product-title} cluster by using the web console.
 * Have access to the cluster as a user with `admin` privileges.
 See xref:../authentication/using-rbac.adoc[Using RBAC to define and apply permissions].
 * Have a recent xref:../backup_and_restore/backing-up-etcd.adoc#backup-etcd[etcd backup] in case your upgrade fails and you must xref:../backup_and_restore/disaster_recovery/scenario-2-restoring-cluster-state.adoc#dr-restoring-cluster-state[restore your cluster to a previous state].
+* If your cluster uses manually maintained credentials, ensure that the CCO is in an upgradeable state. For more information, see _Upgrading clusters with manually maintained credentials_ for xref:../installing/installing_aws/manually-creating-iam.adoc#manually-maintained-credentials-upgrade_manually-creating-iam-aws[AWS], xref:../installing/installing_azure/manually-creating-iam-azure.adoc#manua ...
```
s/CCO/Cloud Credential Operator (CCO)

bmcelvee commented on updating/updating-restricted-network-cluster.adoc:
```diff
@@ -19,6 +19,7 @@ If multiple clusters are present within the restricted network, mirror the requi
 * Have access to the cluster as a user with `admin` privileges.
 See xref:../authentication/using-rbac.adoc[Using RBAC to define and apply permissions].
 * Have a recent xref:../backup_and_restore/backing-up-etcd.adoc#backup-etcd[etcd backup] in case your upgrade fails and you must xref:../backup_and_restore/disaster_recovery/scenario-2-restoring-cluster-state.adoc#dr-restoring-cluster-state[restore your cluster to a previous state].
+* If your cluster uses manually maintained credentials, ensure that the CCO is in an upgradeable state. For more information, see _Upgrading clusters with manually maintained credentials_ for xref:../installing/installing_aws/manually-creating-iam.adoc#manually-maintained-credentials-upgrade_manually-creating-iam-aws[AWS], xref:../installing/installing_azure/manually-creating-iam-azure.adoc#manua ...
```
s/CCO/Cloud Credential Operator (CCO)

bmcelvee commented:
Just a few suggestions, otherwise LGTM!
